boringtun/src/device/mod.rs

```diff
@@ -678,31 +716,43 @@ impl Device {
             }
         });
 
-        let mut buf_count = 0;
-        loop {
-            let mut buf = buf_rx.try_recv().unwrap_or_else(|_| {
-                buf_count += 1;
-                log::info!("buf_count={buf_count}");
-                datagram_buffer()
-            });
-
-            // Read packets from the socket.
-            let (packet_len, addr) = udp.recv_from(&mut buf[..]).await.map_err(|e| {
-                log::error!("UDP recv_from error {e:?}");
-                Error::IoError(e)
-            })?;
-
-            let packet_buf = PacketBuf { packet_len, buf };
-
-            if let Err(mpsc::error::TrySendError::Full((packet_buf, addr))) =
-                dec_tx.try_send((packet_buf, addr))
-            {
-                log::warn!("Back-pressure on dec_tx"); // TODO: remove this log
-                if dec_tx.send((packet_buf, addr)).await.is_err() {
-                    return Ok(()); // Decryption task has been dropped
+        let receive_task = tokio::task::spawn(async move {
+            let mut buf_count = 0;
+            loop {
+                let mut buf = buf_rx.try_recv().unwrap_or_else(|_| {
+                    buf_count += 1;
+                    log::info!("buf_count={buf_count}");
+                    datagram_buffer()
+                });
+                // Read packets from the socket.
+                let (packet_len, addr) = udp.recv_from(&mut buf[..]).await.map_err(|e| {
+                    log::error!("UDP recv_from error {e:?}");
+                    Error::IoError(e)
+                })?;
+
+                let packet_buf = PacketBuf { packet_len, buf };
+
+                // TODO: handle closed
+                if let Err(mpsc::error::TrySendError::Full((packet_buf, addr))) =
+                    dec_tx.try_send((packet_buf, addr))
+                {
+                    log::warn!("Back-pressure on dec_tx incoming"); // TODO: remove this log
+                    if dec_tx.send((packet_buf, addr)).await.is_err() {
+                        return Ok::<(), Error>(()); // Decryption task has been dropped
+                    }
                 }
             }
+        });
+
+        // TODO: abort tasks on drop
```

Comment:
We could use [`CancellationToken`][1]s for this.

[1]: https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html